Init: chromium darwin

michaelCTS opened the pull request:
###### Description of changes

Provides a simple chromium package for darwin as well as a `makeDarwinApp`.

There are still some possible improvements, but I would like to keep this PR small and add those improvements in later PRs instead of having one enormous PR.

 - an update script
 - a common `mkDerivation.meta` (though that may be messy)
 - `makeDarwinApp` should be able to retrieve the version from the source (`Info.plist` should contain a field with the app version)


**Notes**

The diff looks like `chromium/default.nix` was edited, but I moved it all into `chromium/linux` and created a new one. The diff is unfortunately not better, but I hope that won't be a problem

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

closes #247855

Commit:
Init chromium-bin aarch64-darwin + extractDarwinApp

It's not clean, but it works.
What's missing is an update script and how to get the actual version without downloading the entire archive first.

extractDarwinApp is a helper for making derivations from
.app and .dmg bundles.

toonn commented on pkgs/build-support/make-darwin-app/default.nix:
```diff
@@ -0,0 +1,74 @@
+{
```
Maybe we should call this unpack-darwin-app or something? The current name causes a bit of confusion with make-darwin-bundle.

michaelCTS replied:
Renamed to `mkDarwinAppFromBin`. I think that conveys better what's going on.

Decided against unpack as that seems like something exclusively for `unpackPhase`. Also considered `mkDarwinDerivation`, but that's too generic imo.

Also added a description of the function.

michaelCTS replied:
Naming things is hard :smile: I'll go with `extract-darwin-app`.

michaelCTS replied:
That's done now. Thanks for the suggestions!

toonn commented on pkgs/build-support/make-darwin-app/default.nix:
```diff
@@ -0,0 +1,74 @@
+    * These contain multiple folders after extraction
+    * which requires special treatment in the unpackPhase
+    */
+  isDmg ? false,
```
Is there no way to detect this? Maybe always set sourceRoot and then afterwards detect the app is just one level deeper if it's not a dmg?

michaelCTS replied:
That a good suggestion which be an improvement for another PR. I'd like to keep this PR as small a possible without too many possible code paths and thus opportunities for things to go wrong.

michaelCTS replied:
Also done in [53ce772](https://github.com/NixOS/nixpkgs/commit/53ce772072f61b49b7881fa404c5287a8e2ae2d7)

toonn commented on pkgs/build-support/make-darwin-app/default.nix:
```diff
@@ -0,0 +1,74 @@
+
+    # Make application available in bin/
+    if [[ "${toString wrapBinary}" = "1" ]] ; then
+      makeWrapper $binDir/${binaryName} $out/bin/${pname}
```
Do we have any idea why? Is this about the binary not having execute permissions or something?

michaelCTS replied:
It seems to have something to do with the location of the binary. It can't find the framework and libs otherwise. The same goes for firefox, which comes from a `.dmg`. Executing the wrapped binary from any CWD works, but if the binary is moved or linked to (`ln`), it doesn't work.

If someone else wants to invest the time to patch the binary, they can be my guest.

toonn commented on pkgs/applications/networking/browsers/chromium/darwin/default.nix:
```diff
@@ -0,0 +1,41 @@
+{
```
Maybe we should add a chromium-bin instead, mirroring what we have for Firefox?

michaelCTS replied:
For the uninitiated, what do these suffixes mean?

 - `-bin`
 - `-bin-unwrapped`
 - `-unwrapped`

michaelCTS replied:
Nvm, found out https://nixos.wiki/wiki/Package_naming_conventions

toonn commented on pkgs/applications/networking/browsers/chromium/darwin/default.nix:
```diff
@@ -0,0 +1,41 @@
+  };
+
+  wrapBinary = true;
+  packageMeta = {
```
Maybe we can set `hydraPlatforms = []`? Since we're just fetching an archive and unpacking it, having Hydra build and cache it isn't very useful.

toonn replied:
firefox-bin also does this FWIW.

michaelCTS replied:
Done in [0ef984b](https://github.com/NixOS/nixpkgs/commit/0ef984bc970687f8df65205a836924c099d61906). Put it in `mkDarwinAppFromBin`

toonn reviewed (commented):
Would be nice if we could get a Linux maintainer's opinion on the reorganizing of the package.

amjoseph-nixpkgs commented:
> Would be nice if we could get a Linux maintainer's opinion on the reorganizing of the package.

Yeah that reorg is going to create a massive merge conflict with pending PRs like #229265.

amjoseph-nixpkgs commented on pkgs/build-support/make-darwin-app/default.nix:
```diff
@@ -0,0 +1,74 @@
+    meta = {
+      platforms = lib.platforms.darwin;
+      mainProgram = binaryName;
+      name = pname;
```
Needs `meta.sourceProvenance`, since this isn't built from source.

michaelCTS replied:
Done in [0ef984b](https://github.com/NixOS/nixpkgs/commit/0ef984bc970687f8df65205a836924c099d61906)

amjoseph-nixpkgs reviewed (dismissed):
Since `chromium` is built from source and this isn't, they zero Nix code; there's really no benefit to having them be the same package.  @toon had [the right idea](https://github.com/NixOS/nixpkgs/pull/244152/files#r1267084736); rename this to `chromium-bin` and put an `if stdenv.hostPlatform.isDarwin then chromium-bin else ...` in `pkgs/top-level/all-packages.nix`.

michaelCTS commented:
@amjoseph-nixpkgs there doesn't seem to be an existing chromium-bin. There wouldn't be a need for `if stdenv.hostPlatform.isDarwin then chromium-bin else ...` in `all-packages.nix`, correct?

amjoseph-nixpkgs commented:
> @amjoseph-nixpkgs there doesn't seem to be an existing chromium-bin.

Right; I was suggesting that you create one.  But, as you point out:

> There wouldn't be a need for `if stdenv.hostPlatform.isDarwin then chromium-bin else ...` in `all-packages.nix`, correct?

I guess if you really want to avoid creating another top-level entry you could just have `callPackage ../applications/networking/browsers/chromium/darwin.nix {}`.

The `-bin` convention is pretty well-established, though.  We have 163 packages in `all-packages.nix` with the `-bin` suffix (including `firefox`).

amjoseph-nixpkgs commented:
Hey, looking much better, thanks a bunch for reorganizing this.  I'll leave the rest of the review up to people who have Darwin hardware to test on.  Nice work!

toonn commented on pkgs/applications/networking/browsers/chromium-bin/default.nix:
```diff
@@ -0,0 +1,8 @@
+}:
+if stdenv.isDarwin
+then callPackage ./darwin.nix {}
+else throw "chromium-bin has not been packaged for linux. use chromium instead"
```
This deviates from the common pattern of listing supported platforms explicitly in `meta.platforms`. I think that's better than going with a custom solution.

michaelCTS replied:
I'm not sure what you mean. Could you point me to an example?

toonn replied:
Most any package expression will do as an example. It's just either not including a platform in `meta.platforms` or including it but adding it to `meta.broken` too.

michaelCTS replied:
Oh, I see. The reason for doing it this way is because of the `extractDarwinApp`. Were a linux version to be introduced, it would require creating a new derivation, and IMO it would be better for that to be a separate file.
I imagine that would require a common meta attrset.

We can let the linux packager find out how to handle that. I'll just return the darwin package for now.

michaelCTS replied:
Done in 5f5e9600e882b8540d4cc7edbdb7403692bde909

toonn commented on pkgs/build-support/extract-darwin-app/default.nix:
```diff
@@ -0,0 +1,85 @@
+
+/** A function to make a derivation from a darwin binary
+  *
+  * It makes the assumption that the binary comes in a ".app" or ".dmg".
```
Shouldn't `.app` here be `.zip`?

michaelCTS replied:
https://github.com/NixOS/nixpkgs/pull/244152/commits/2f85c6eed236dd0c1e523414630a6f6c481e46c7 updates the doc

toonn commented on pkgs/build-support/extract-darwin-app/default.nix:
```diff
@@ -0,0 +1,85 @@
+
+    mkdir -p $out/bin
+    if [[ "${toString isDmg}" = "1" ]] ; then
+      # .dmg uses sourceRoot, which means PWD is already in the sourceRoot
```
Can't we avoid setting the `sourceRoot` to inside the `.app` bundle so that we can do away with these branches and just copy the bundle in both cases?

michaelCTS replied:
[2f85c6e](https://github.com/NixOS/nixpkgs/pull/244152/commits/2f85c6eed236dd0c1e523414630a6f6c481e46c7) treats all the usecases above but with the existing branches. I couldn't conceive a solution without branches.

toonn replied:
Just setting `sourceRoot = "."` should work. It only makes passing an app as the source slightly harder but I don't think that'd be a common case?

michaelCTS replied:
It does seem like a valid usecase. Is it a problem that it's handled?

toonn replied:
It makes everything more complex for very little gain though. When would you want to use a straight app as source and be unable to use the directory containing it instead?

toonn replied:
@michaelCTS, do you disagree strongly with the simplification?

toonn reviewed (commented):
Much simpler this way, I think making a new package was for the best.

Commit:
extractDarwinApp: support more source formats

The arguments and documentation implied that only `.app` and `.dmg`
we supported, but that wasn't true.
The case for `src = ./some/folder` and non-dmg archives with multiple
folders was not supported.

Now, the user can specify srcHasManySubfolders which sets the
`sourceRoot` for `mkDerivation` to a default value.
The user also can directly set `sourceRoot` themselves.

Commit:
refactor(chromium-bin) Only return darwin package

This allows nix to handle the supported platforms
instead of throwing a string for unsupported platforms.

jakuzure commented:
> Hey, looking much better, thanks a bunch for reorganizing this. I'll leave the rest of the review up to people who have Darwin hardware to test on. Nice work!

I just tried it out via home-manager (set the package option to chromium-bin) and it seems to be working fine. I couldn't get the extensions option to work though, but that might be related to the bin maybe? Installing them directly in the browser worked fine

michaelCTS commented:
@WeetHet as per my [last comment](https://github.com/NixOS/nixpkgs/pull/244152#discussion_r1327378209), there's no time for this on my end. I'm actually going to sell my mac as it's not necessary anymore.

Feel free to branch off my branch and make the changes @toonn wanted (that I disagreed with). It's probably what will get the PR approved.

lrworth commented:
I've opened #324701, simplifying this as much as possible in order to get it merged. Someone with more motivation may get something like `extract-darwin-app` working in the future.
